### PR TITLE
Add codes to work with brain data and add slice loader

### PR DESCRIPTION
**Pull request by chaithyagr**
This PR handles:
1) Obtaining data from brain dataset. Earlier there were some flow issues. Also it didnt make sense to save shapes just because it is brain data, especially after we have cropped the data when we request for complex image.
2) We also added codes to split up data from different slices as different data points, allowing us to go to variable batch sizes as needed in future.

### CODE REVIEW

**Comment by chaithyagr** [comment]
I have implemented some basic stuff and tested it locally. @zaccharieramzi we would ideally need to setup a complete test workflow to test all parts

**Comment by zaccharieramzi** [comment]
Hmmm

1. well the idea was that the only time when there is an uncertainty about the expected shape of the image is for the brain data. In the case of the knee data, you know that the expected ground truth image (cropped) is going to have a `320 x 320` shape. For brain data, it is not known in advance. Also, even when we use complex images, we know that the shape is going to be `320 x 320` (it is hardcoded). Therefore I am not sure why we would want to have the shape in other cases. 
1 (bis). I think the issues with the brain data are because you tried to use the noisy datasets with brain data (or multicoil data). Indeed they were not planned for this so that's why you had this problem. Just to be clear, these datasets shouldn't be used for MRI reconstruction, they were for other projects (like denoising score matching). I am not sure it would make sense to use it for multicoil data, and therefore brain data (did you?).
2. I understand what you did, but I am not sure why it helps going to variable batch sizes in the future. I also am not sure in the current state why it's better than `slice_random`.

I think I would need to add in some docs to make those points clearer anyway.
But please feel free to discuss them.

The test workflow is already present, it's just that it was in travis which is not working anymore in free-mode.
I will make a PR to setup GitHub actions in a minute.

**Comment by chaithyagr** [comment]
> I understand what you did, but I am not sure why it helps going to variable batch sizes in the future. I also am not sure in the current state why it's better than slice_random.

Wouldnt `slice_random` just select one random slice from each dataset?

**Comment by zaccharieramzi** [comment]
> Wouldnt slice_random just select one random slice from each dataset?

Yes but a different one in each epoch, so in the end, if you do sufficiently many epochs you will have seen all the slices.
In other words, the random selection is done on-the-fly.

**Comment by chaithyagr** [comment]
Hmm yes, but in that case epoch isnt exactly as defined right? I mean when you do 1 epoch in this case, you dont visit all of the data. Further, in case we plan to use conventional schemes to carry out reconstruction, having this could be helpful right?

**Comment by zaccharieramzi** [comment]
I mean there is not one definition of epoch. I defined it this way:
"an epoch corresponds to selecting one slice, selected at random on-the-fly, of each volume in the dataset".

If we want to carry out conventional reconstruction schemes, then we would certainly load the whole data for one volume at once, without slicing. Also we would want to have the same mask for all the slices in the volume, which would be more difficult to ensure if the slices were separate elements in the dataset.

**Comment by chaithyagr** [comment]
I think i underrstand your points. In this case, let me revert to only changes as needed and add some tests.

**Comment by chaithyagr** [comment]
I seem to be back at this, I really need to use a way to have data split based on slices.

My usecase for `split_slices` is that I want to obtain all the spectrums of all the fastMRI data.
I would want to have kspaces for all the slices and carry out something non DL related to it (like learning a target density).

While I can write my own function as it is very specific to me, I felt that this would just make the library more flexible. Is there any downside to this?

What I am looking for is kinda exactly this which is on official repo : https://github.com/facebookresearch/fastMRI/blob/88e7b6e363485cea90873a70fb664f60be1d9b47/fastmri/data/mri_data.py#L172

File: tf_fastmri_data/dataset_builder.py [hunk 1]
```diff
@@ -118,18 +137,13 @@ def _build_datasets(self):
...
             )
-        if self.brain:
+        if self.save_shapes:
             output_shape_ds = tf.data.Dataset.from_tensor_slices(
                 [load_output_shape_from_file(f) for f in self.filtered_files],
             )
             self._raw_ds = tf.data.Dataset.zip(
                 (self._raw_ds, output_shape_ds)
             )
-            self._raw_ds = self._raw_ds.map(
```
Comment (zaccharieramzi) [inline tf_fastmri_data/dataset_builder.py:128]: Why did you remove this?

File: tf_fastmri_data/dataset_builder.py [hunk 2]
```diff
@@ -99,10 +114,14 @@ def _get_path_default(self,):
...
     def _build_datasets(self):
+        if self.split_slices:
+            self.files_ds = self.files_ds.map(
+                lambda x : (x[0], int(x[1]))
+            )
         self._raw_ds = self.files_ds.map(
             partial(
                 load_data_from_file,
-                slice_random=self.slice_random,
+                slice_random=self.slice_random or self.split_slices,
```
Comment (zaccharieramzi) [inline tf_fastmri_data/dataset_builder.py:124]: I think we should then change the name of this argument, because it does not reflect what's actually going on.
Maybe just another argument would be nice.

File: tf_fastmri_data/dataset_builder.py [hunk 3]
```diff
@@ -18,12 +18,14 @@ def __init__(
...
             slice_random=False,
             contrast=None,
+            split_slices=False,
             af=4,
             shuffle=False,
             seed=0,
             prebuild=True,
             repeat=True,
             n_samples=None,
+            save_shapes=False,
```
Comment (zaccharieramzi) [inline tf_fastmri_data/dataset_builder.py:28]: it's not really saving, maybe `output_shapes` would be better suited.

File: tf_fastmri_data/datasets/noisy.py [hunk 4]
```diff
@@ -31,7 +31,7 @@ def __init__(
         if self.mode == 'test':
             raise NotImplementedError('Noisy dataset only works for train/val')
 
-    def _preprocessing_train(self, image):
+    def _preprocessing_train(self, image, *args):
```
Comment (zaccharieramzi) [inline tf_fastmri_data/datasets/noisy.py:34]: I think this file shouldn't be affected by your change. Why is it?

File: tf_fastmri_data/h5.py [hunk 5]
```diff
@@ -4,15 +4,15 @@
 import tensorflow_io as tfio
 
 
-def load_data_from_file(fpath, slice_random=False, no_kspace=False, multicoil=False, mode='train'):
+def load_data_from_file(fpath, slice_num=-1, slice_random=False, no_kspace=False, multicoil=False, mode='train'):
```
Comment (zaccharieramzi) [inline tf_fastmri_data/h5.py:7]: let's not call it `slice_num` because you in `load_metadata_from_file` you use the same var name for a different context.
maybe `slice_index` instead.

File: tf_fastmri_data/dataset_builder.py (hunk 2 above)
Comment (zaccharieramzi) [reply tf_fastmri_data/dataset_builder.py:124]: Actually you don't even need this argument, you can always use the fact that `slice_index > -1` instead at run time.

File: tf_fastmri_data/h5.py [hunk 6]
```diff
@@ -50,8 +50,11 @@ def load_data_from_file(fpath, slice_random=False, no_kspace=False, multicoil=Fa
     else:
         slices = (0, n_slices)
     if mode == 'train':
-        image = h5_tensors(image_name)[slices[0]:slices[1]]
-        if slice_random:
+        if slice_num > -1:
+            image = h5_tensors(image_name)[slice_num:slice_num+1]
```
Comment (zaccharieramzi) [inline tf_fastmri_data/h5.py:54]: I think you are better off using another condition when defining `slices`. Like:
```python
elif slice_index >-1:
	slices = (slice_index, slice_index + 1)
```

This way the rest of the implem can remain unchanged.

**Review by zaccharieramzi — CHANGES_REQUESTED** [review]
I agree with the proposed idea of splitting slices.
Just a few comments on the current implementation which works but can be confusing for future developments imho.

File: tf_fastmri_data/datasets/noisy.py (hunk 4 above)
Comment (chaithyagr) [reply tf_fastmri_data/datasets/noisy.py:34]: Ahh this was actually with respect to getting codes to work as well. There were some extra args being sent out (if i am right shapes / kspaces, not sure at the moment. This would fail here for brain data.

File: tf_fastmri_data/h5.py (hunk 5 above)
Comment (chaithyagr) [reply tf_fastmri_data/h5.py:7]: Ahh yes

File: tf_fastmri_data/h5.py (hunk 6 above)
Comment (chaithyagr) [reply tf_fastmri_data/h5.py:54]: Ahh yes..

File: tf_fastmri_data/datasets/noisy.py (hunk 4 above)
Comment (zaccharieramzi) [reply tf_fastmri_data/datasets/noisy.py:34]: yup yup but for now let's try to keep this PR concise with only the slicing

File: tf_fastmri_data/dataset_builder.py (hunk 1 above)
Comment (chaithyagr) [reply tf_fastmri_data/dataset_builder.py:128]: Oh this code was broken, I actually wanted to ask you this. This code from what i see tries to open dataset so that we have it as (kspace, image, shape) right? That is not happening. In fact the code errors out right here. 
I think i am missing something with respect to the aim of this dataset map.

File: tf_fastmri_data/datasets/noisy.py (hunk 4 above)
Comment (chaithyagr) [reply tf_fastmri_data/datasets/noisy.py:34]: Yes on it

File: tf_fastmri_data/h5.py (hunk 5 above)
Comment (chaithyagr) [reply tf_fastmri_data/h5.py:7]: Also, this would mean we would have to encode select random slice under `slice_index=-2` say. We ok with that i hope?

File: tf_fastmri_data/dataset_builder.py (hunk 2 above)
Comment (chaithyagr) [reply tf_fastmri_data/dataset_builder.py:124]: Wait a second, we need to tell early on that we are taking slices, so that there is no parsing issues. The code otherwise early on forms graph assuming that we may or may not take with slices, due to which the shape is drastically changed. I exactly faced this, hence I resorted to sending that slice_random is `True`:

```
TypeError: 'kspace_shape' must have the same nested structure in the main and else branches:
    
    The two structures don't have the same nested structure.
    
    First structure: type=tuple str=(None, None, None)
    
    Second structure: type=tuple str=(None, None, None, None)
    
```


What I propose is to rather use different argument or just call this argument as `select_slices`, which means we select slices, and if `slice_index = -1`, we choose at random. What do you think?

**Review by chaithyagr — COMMENTED** [review]
A few extra comments

**Review by zaccharieramzi — CHANGES_REQUESTED** [review]
one leftover and a small comment

Also trying to see what happened in the map you removed.

File: tf_fastmri_data/dataset_builder.py [hunk 7]
```diff
@@ -18,12 +18,14 @@ def __init__(
...
             slice_random=False,
             contrast=None,
+            split_slices=False,
             af=4,
             shuffle=False,
             seed=0,
             prebuild=True,
             repeat=True,
             n_samples=None,
+            output_shapes=False,
```
Comment (zaccharieramzi) [inline tf_fastmri_data/dataset_builder.py:28]: actually, let's make this default to None and set it to different default values depending on the value of brain (we should do the same with multicoil tbh but we can do it in a different PR).
```suggestion
            output_shapes=None,
```

File: tf_fastmri_data/dataset_builder.py [hunk 8]
```diff
@@ -33,6 +35,7 @@ def __init__(
         self.dataset = dataset
         self._check_dataset()
         self.brain = brain
+        self.output_shapes = output_shapes
```
Comment (zaccharieramzi) [inline tf_fastmri_data/dataset_builder.py:38]:
```suggestion
        self.output_shapes = self.brain if output_shapes is None else output_shapes
```

File: tf_fastmri_data/dataset_builder.py (hunk 1 above)
Comment (zaccharieramzi) [reply tf_fastmri_data/dataset_builder.py:128]: Can you tell me exactly the code you ran to see a failure?
I tried the following:

```python
from tf_fastmri_data.datasets.cartesian import CartesianFastMRIDatasetBuilder

ds_builder = CartesianFastMRIDatasetBuilder(
    dataset='val',
    brain=True,
    multicoil=True,
)

ds = ds_builder.preprocessed_ds

res = next(iter(ds))
```

without errors

File: tf_fastmri_data/datasets/noisy.py (hunk 4 above)
Comment (zaccharieramzi) [reply tf_fastmri_data/datasets/noisy.py]: I think you misunderstood me, I meant to remove the changes in the noisy dataset altogether so that we keep the PR concise (not this line in particuar)

Comment (chaithyagr) [reply tf_fastmri_data/datasets/noisy.py]: Ahhh, thing is I dont remember the changes, I will compare and revert accordingly.. xD

**Comment by chaithyagr** [comment]
The codes I added in nosiy.py are needed to have correct dimensions while adding noise. However, I dont think it covers all cases right, I am still debugging it right. But it covers the obvious cases for now.

**Comment by zaccharieramzi** [comment]
I understand perfectly what you did but I want to make this PR focused on just the slices, so we can have another discussion about the multicoil noise in another one.

**Comment by chaithyagr** [comment]
> I understand perfectly what you did but I want to make this PR focused on just the slices, so we can have another discussion about the multicoil noise in another one.

Makes sense. I am reverting and perhaps I will have a look at why the tests are failing if they do.

File: tf_fastmri_data/dataset_builder.py (hunk 1 above)
Comment (chaithyagr) [reply tf_fastmri_data/dataset_builder.py:128]: Lol, I think I messed my environments, This is only causing issues for me right now.. Reverting this

**Review by zaccharieramzi — APPROVED** [review]

**Comment by zaccharieramzi** [comment]
LGTM, merging when green

File: tf_fastmri_data/dataset_builder.py [hunk 9]
```diff
@@ -157,10 +176,19 @@ def preprocessed_ds(self):
             self._build_datasets()
         return self._preprocessed_ds
 
+    def _preprocessing_train(self, image, *args):
```
Comment (chaithyagr) [inline tf_fastmri_data/dataset_builder.py:179]: Ahh I wanted to ask this to you @zaccharieramzi , I have added a method to just return kspace and images so that we can build on top of this further, or just pickup this raw data (for conventional reconstructions)..
This will remove the assert Not implemented test.. 
We okay with that right?

Comment (zaccharieramzi) [reply tf_fastmri_data/dataset_builder.py:179]: sure
